docs/typescript.md

```diff
@@ -1,14 +1,590 @@
+- For type assertions to an incompatible shape, use `as unknown as` as a last resort rather than `any` or `as any`.
+
+##### `as` is always acceptable to use for TypeScript syntax other than type assertion
+
+- Writing type guards often requires using the `as` keyword.
+
+```typescript
+function isFish(pet: Fish | Bird): pet is Fish {
+  return (pet as Fish).swim !== undefined;
+}
+```
+
+- Key remapping in mapped types uses the `as` keyword.
+
+```typescript
+type MappedTypeWithNewProperties<Type> = {
+  [Properties in keyof Type as NewKeyType]: Type[Properties];
+};
+```
+
+##### `as` is acceptable to use for typing data objects whose shape and contents are determined at runtime, externally, or through deserialization
+
+Preferably, this typing should be accompanied by runtime schema validation performed with type guards and unit tests.
+
+- e.g. The output of `JSON.parse()` or `await response.json()` for a known JSON input.
+- e.g. The type of a JSON file.
+
+<!-- TODO: Add example -->
+
+##### `as` may be acceptable to use in tests, for mocking or to exclude irrelevant but required properties from an input object
+
+It's recommended to provide accurate typing if there's any chance that omitting properties affects the accuracy of the test.
+
+If that is not the case, however, mocking only the properties needed in the test improves readability. It makes the intention and scope of the mock clear, and is more convenient to write, which can encourage test writing and improve test coverage.
+
+<!-- TODO: Add examples -->
+
+### Compiler Directives
+
+TypeScript provides several directive comments that can be used to suppress TypeScript compiler errors. Using these to ignore typing issues is dangerous and reduces the overall effectiveness of TypeScript.
+
+#### Acceptable usages of `@ts-expect-error`
+
+##### Use `@ts-expect-error` to force runtime execution of a branch for validation or testing
+
+Sometimes, there is a need to force a branch to execute at runtime for security or testing purposes, when that branch has correctly been inferred as being inaccessible by the TypeScript compiler.
+
+
+✅
+
+> **Error:** This comparison appears to be unintentional because the types '`0x${string}`' and '"__proto__"' have no overlap.ts(2367)
+
+```typescript
+exampleFunction(chainId: `0x${string}`) {
+    // @ts-expect-error Suppressing to perform runtime check
+    if (chainId === '__proto__') {
+      return;
+    }
+    ...
+}
+```
+
+##### `@ts-expect-error` may be acceptable to use in tests, to intentionally break features
+
+✅
+
+```typescript
+// @ts-expect-error Suppressing to test runtime error handling
+
+// @ts-expect-error Intentionally testing invalid state
+
+// @ts-expect-error We are intentionally passing bad input.
+```
+
+##### If accompanied by a TODO comment, `@ts-expect-error` is acceptable to use for marking errors that have clear plans of being resolved
+
+<!-- TODO: Add example -->
+
+#### Avoid `any`
```

Comment:
It looks like this is currently in the "Compiler Directives" section. Was this meant to be its own section?

Reply:
Designating `any` as a compiler directive was intentional, just to highlight that it's not an actual type. 

The hierarchy of headings is a little iffy. I currently have "Avoid something" and "Acceptable usages of something" consistently at the same level (h4), because pushing down the actual entries (h6) an additional level would make them too small to read.

Reply:
Hmm, I had never heard of `any` being referred to as a "compiler directive". I took "compiler directive" to mean the triple-slash directives. Even if this is technically true on some level, it doesn't seem like the most intuitive way to organize the document.

Reply:
Hmm I can see that "compiler directives" could be confusing. What do you think about naming the last section "Escape Hatches" instead? That could cover `@ts-expect-error`, `@ts-ignore`, `any`, in order of preference. 

The `@ts-ignore` section could only include a brief paragraph comparing it to `any`: The two are functionally identical, except `any` is much worse, because it's like applying `@ts-ignore` to every occurrence of the assignee variable throughout the codebase and even in downstream modules. That and the fact that `@ts-ignore` applies to entire lines that could contain multiple unrelated errors are the only differences.

I think this would provide better context for the `any` section that follows so it doesn't feel out of place.

Reply:
"Escape hatches" sounds like a great description of what these are

Reply:
Applied here: https://github.com/MetaMask/contributor-docs/pull/74/commits/6f229675d64adb4fcc2452c89c5b86c7f4a1f604

Reply:
I like this as well!